### ISSUE TITLE
Filtering an enter() selection results in odd behaviour

### ISSUE THREAD

**Issue by rivo** [d3/d3-selection#134]
I'm running `filter()` on an `enter()` selection because I only want specific elements to be created, not all missing ones. Unfortunately, when I then merge that filtered selection into the existing element selection, I only get the newly created element. Here's some code that illustrates this:

```js
var arr = ["one","two","three"],
  ul = d3.select("ul"),
  round = 1;
  
add("two");
add("three");
  
function add(selection) {
  li = ul.selectAll("li").data(arr);
  li.exit().remove();
  li = li.enter()
    .filter(function(d) { return d === selection; })
    .append("xhtml:li")
    .merge(li);
  li.text(function(d) { return "Round: " + round + ", " + d; });
  round++;
}
```
https://jsfiddle.net/yzjhx06r/

My understanding is that the `li` variable in line 15 should contain all existing "li" elements. So the resulting text should be:

- Round: 2, two
- Round: 2, three

Instead, it is:

- Round: 1, two
- Round: 2, three

So the merge (line 14) does not work as expected. Is there a misunderstanding on my side here? Or could this be a bug?

**Comment by mbostock** [comment]
This is the expected behavior. [*selection*.merge](https://github.com/d3/d3-selection/blob/master/README.md#selection_merge) requires the two selections to have the same shape: “Any missing (null) elements in this selection are filled with the corresponding element, if present (not null), from the specified selection.” But [*selection*.filter](https://github.com/d3/d3-selection/blob/master/README.md#selection_filter) does not preserve the shape of the selection, as it drops any elements for which the filter function returns false.

Per the README, if you want to preserve the shape of the selection when filtering, you need to use [*selection*.select](https://github.com/d3/d3-selection/blob/master/README.md#selection_select) instead of *selection*.filter. So, instead of:

```js
var foo = selection.filter(function(d) { return d.type === "foo"; });
 ```

You would say:

```js
var foo = selection.select(function(d) { return d.type === "foo" ? this : null; });
 ```

In the latter case `foo` has the same shape as `selection`, and thus can be merged.

There is an open feature request #89 for generalizing *selection*.merge, but no one’s thought of an acceptable general-purpose merge yet, so for now we’re sticking with the simplest one that works for the data-join.

I’m not entirely sure what you are trying to accomplish based on the snippet you posted, but the other way you might tackle this problem is to filter your data *before* computing the data-join. You might also want to use a key function when computing the data-join.

If you have further questions, please consider using [Stack Overflow tag d3.js](https://stackoverflow.com/tags/d3.js) to ask for help. Although I make an effort to assist everyone that asks, I am not always available to provide help promptly or directly. Stack Overflow provides a better collaborative forum for self-help: tens of thousands of D3-related questions have already been asked there, and some answered questions may be relevant to you.

If you have a question about D3’s behavior and want to discuss it with other users, also consider the [d3-js Google Group](https://groups.google.com/forum/#!forum/d3-js) or [joining](https://d3-slackin.herokuapp.com/) the [d3-js Slack](https://d3js.slack.com/).

Thank you! 🤗

**Comment by rivo** [comment]
I posted this here (instead of SO) because it felt like a bug. :-) Nonetheless, I really appreciate your help there! I'm not sure the meaning of "shape" is immediately obvious without knowing `selection.merge`'s internal implementation details. But I will definitely be careful now when using `filter` in conjunction with `merge`.

Again, thanks!

ps. The example I gave was for lazy initialization: "Add this one DOM element if it's missing."
